validators/valset/snapshot/set.go

```diff
@@ -0,0 +1,538 @@
+		})
+	}
+
+	// check the tally for the proposed validator
+	tally := snap.Count(func(v *valset.Vote) bool {
+		return v.Candidate.Equal(candidate)
+	})
+
+	// If more than a half of all validators voted
+	if tally > snap.Set.Len()/2 {
+		if authorize {
+			// add the candidate to the validators list
+			if err := snap.Set.Add(candidate); err != nil {
+				return err
+			}
+		} else {
+			// remove the candidate from the validators list
+			if err := snap.Set.Del(candidate); err != nil {
+				return err
+			}
+
+			// remove any votes casted by the removed validator
+			snap.RemoveVotes(func(v *valset.Vote) bool {
+				return v.Validator == candidate.Addr()
+			})
+		}
+
+		// remove all the votes that promoted this validator
+		snap.RemoveVotes(func(v *valset.Vote) bool {
+			return v.Candidate.Equal(candidate)
+		})
+	}
+
+	saveSnap()
+	s.store.updateLastBlock(header.Number)
+
+	return nil
+}
+
+func (s *SnapshotValidatorSet) Propose(candidate validators.Validator, auth bool, proposer types.Address) error {
+	s.candidatesLock.Lock()
+	defer s.candidatesLock.Unlock()
+
+	candidateAddr := candidate.Addr()
+
+	for _, c := range s.candidates {
+		if c.Validator.Addr() == candidateAddr {
+			return ErrAlreadyCandidate
+		}
+	}
+
+	snap := s.getLatestSnapshot()
+	if snap == nil {
+		return ErrSnapshotNotFound
+	}
+
+	included := snap.Set.Includes(candidateAddr)
+
+	// safe checks
+	if auth && included {
+		return ErrCandidateIsValidator
+	} else if !auth && !included {
+		return ErrCandidateNotExistInSet
+	}
+
+	// check if we have already voted for this candidate
+	count := snap.Count(func(v *valset.Vote) bool {
+		return v.Candidate.Addr() == candidateAddr && v.Validator == proposer
+	})
+
+	if count == 1 {
+		return ErrAlreadyVoted
+	}
+
+	if auth {
+		s.candidates = append(s.candidates, &valset.Candidate{
+			Validator: candidate,
+			Authorize: auth,
+		})
+	} else {
+		// get candidate validator information from set
+		// because don't want user to specify data except for address
+		// in case of removal
+		validatorIndex := snap.Set.Index(candidate.Addr())
+		validatorInSet := snap.Set.At(uint64(validatorIndex))
+
+		s.candidates = append(s.candidates, &valset.Candidate{
+			Validator: validatorInSet,
+			Authorize: auth,
+		})
+	}
+
+	return nil
+}
+
+// Votes returns the votes in the snapshot at the specified height
+func (s *SnapshotValidatorSet) Votes(height uint64) ([]*valset.Vote, error) {
+	snapshot := s.getSnapshot(height)
+	if snapshot == nil {
+		return nil, ErrSnapshotNotFound
+	}
+
+	return snapshot.Votes, nil
+}
+
+// Candidates returns the current candidates
+func (s *SnapshotValidatorSet) Candidates() []*valset.Candidate {
+	return s.candidates
+}
+
+// addHeaderSnap creates the initial snapshot, and adds it to the snapshot store
+func (s *SnapshotValidatorSet) addHeaderSnap(header *types.Header) error {
+	// Genesis header needs to be set by hand, all the other
+	// snapshots are set as part of processHeaders
```

Comment:
Why is this comment relevant?

Reply:
Fixed at c14de24e7aff9d30ebb05d813044b1b018d42167